### PR TITLE
Fix closure in closure parsing

### LINKED ISSUE #2656: Cannot parse closure in closure without space
A closure returning a closure without a space cannot be parsed by GCCRS.

I tried this code:

```rust
fn main() {
    let f = |_||x, y| x+y;
    assert_eq!(f(())(1, 2), 3);
}
```

:warning: The default rust formatter will change the code and introduce a space between the pipes, circumventing the bug (the code with the space compiles correctly).

I expected to see this happen: Compile correctly.

Instead, this happened:

```
./test.rs:2:15: error: expecting ‘|’ but ‘||’ found
    2 |     let f = |_||x, y| x+y;
      |               ^~
./test.rs:2:15: error: failed to parse expression in let statement
./test.rs:2:5: error: failed to parse statement or expression in block expression
    2 |     let f = |_||x, y| x+y;
      |     ^~~
./test.rs:4:1: error: unrecognised token ‘}’ for start of item
    4 | }
      | ^
./test.rs:4:1: error: failed to parse item in crate
```

### Meta

- c7728c8020334fecb47789163ba94326aab6208c

### PR DESCRIPTION
**Pull request by P-E-P**
Fix a the parsing of a closure in another closure without a space lead to the lexer creating an OR token.

Fixes #2656

### CODE REVIEW

File: gcc/rust/parse/rust-parse-impl.h [hunk 1]
```diff
@@ -14445,12 +14445,21 @@ Parser<ManagedTokenSource>::parse_closure_expr_pratt (const_TokenPtr tok,
 
 	    if (lexer.peek_token ()->get_id () != COMMA)
 	      {
+		if (lexer.peek_token ()->get_id () == OR)
+		  {
+		    lexer.split_current_token (PIPE, PIPE);
+		  }
```
Comment (CohenArthur) [inline gcc/rust/parse/rust-parse-impl.h:14451]:
```suggestion
		    lexer.split_current_token (PIPE, PIPE);
```
nit

File: gcc/rust/parse/rust-parse-impl.h [hunk 2]
```diff
@@ -14445,12 +14445,21 @@ Parser<ManagedTokenSource>::parse_closure_expr_pratt (const_TokenPtr tok,
...
 		// not an error but means param list is done
 		break;
 	      }
 	    // skip comma
 	    lexer.skip_token ();
 
+	    if (lexer.peek_token ()->get_id () == OR)
+	      {
+		lexer.split_current_token (PIPE, PIPE);
+	      }
```
Comment (CohenArthur) [inline gcc/rust/parse/rust-parse-impl.h:14461]: same here

**Review by CohenArthur — APPROVED** [review]

**Comment by CohenArthur** [comment]
the test is a good one haha

File: gcc/rust/parse/rust-parse-impl.h (hunk 1 above)
Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h]: Fixed.

File: gcc/rust/parse/rust-parse-impl.h (hunk 2 above)
Comment (P-E-P) [reply gcc/rust/parse/rust-parse-impl.h]: Fixed.

**Pull request merged**